be2-scala/src/test/scala/ch/epfl/pop/json/MessageDataProtocolSuite.scala

```diff
@@ -230,4 +232,59 @@ class MessageDataProtocolSuite extends FunSuite with Matchers {
     greetLaoFromExample should equal(expectedGreetLao)
     greetLaoFromBuiltJson should equal(expectedGreetLao)
   }
+
+  test("Parser correctly encodes and decodes ObjectType and rejects incorrect type") {
+    ObjectType.values.foreach(obj => {
+      val fromJson = MessageDataProtocol.objectTypeFormat.write(obj)
+      obj match {
+        case ObjectType.INVALID => assertThrows[IllegalArgumentException] {
+            MessageDataProtocol.objectTypeFormat.read(fromJson)
+          }
+        case _ =>
+          val toType = MessageDataProtocol.objectTypeFormat.read(fromJson)
+          toType shouldBe a[ObjectType]
+      }
+    })
+    val invalidJson = """{"object": "stellarobject"}""".parseJson
+    assertThrows[IllegalArgumentException] {
+      MessageDataProtocol.objectTypeFormat.read(invalidJson)
+    }
+  }
+
+  test("Parser correctly encodes and decodes ActionType and rejects incorrect type") {
+    ActionType.values.foreach(obj => {
+      val fromJson = MessageDataProtocol.actionTypeFormat.write(obj)
+      obj match {
+        case ActionType.INVALID => assertThrows[IllegalArgumentException] {
+            MessageDataProtocol.actionTypeFormat.read(fromJson)
+          }
+        case _ =>
+          val toType = MessageDataProtocol.actionTypeFormat.read(fromJson)
+          toType shouldBe a[ActionType]
+      }
+    })
+    val invalidJson = """{"action": "stellaraction"}""".parseJson
+    assertThrows[IllegalArgumentException] {
+      MessageDataProtocol.actionTypeFormat.read(invalidJson)
+    }
+  }
+
+  test("Parser correctly encodes and decodes VersionType and rejects incorrect type") {
+    VersionType.values.foreach(obj => {
+      val fromJson = MessageDataProtocol.versionTypeFormat.write(obj)
+      obj match {
+        case VersionType.INVALID => assertThrows[IllegalArgumentException] {
+            MessageDataProtocol.versionTypeFormat.read(fromJson)
+          }
+        case _ =>
+          val toType = MessageDataProtocol.versionTypeFormat.read(fromJson)
+          toType shouldBe a[VersionType]
+      }
+    })
+    val invalidJson = """{"version": "stellarversion"}""".parseJson
+    assertThrows[IllegalArgumentException] {
+      MessageDataProtocol.versionTypeFormat.read(invalidJson)
+    }
+  }
```

Comment:
I think it would be nice to separate the two (valid and invalid) case so its easier to see what failed